Inferences/GeneralInduction.cpp

```diff
@@ -0,0 +1,517 @@
+  for (const auto& kv : sch.inductionTerms()) {
+    if (skolem(kv.first)) {
+      return false;
+    }
+  }
+  // check already existing pattern for main literal
+  if (!_done.find(res.first)) {
+    return false;
+  }
+  auto s = _done.get(res.first);
+  // check if the sides for the new pattern are included in the existing one
+  if (includes(s.begin(), s.end(), res.second.begin(), res.second.end())) {
+    if (env.options->showInduction()) {
+      env.beginOutput();
+      env.out() << "[Induction] already inducted on " << *mainLit << " in " << *res.first << " form" << endl;
+      env.endOutput();
+    }
+    return true;
+  }
+  return false;
+}
+
+inline bool sideLitCondition(Literal* main, Clause* mainCl, Literal* side, Clause* sideCl) {
+  auto mainSk = InductionHelper::collectSkolems(main, mainCl);
+  auto sideSk = InductionHelper::collectSkolems(side, sideCl);
+  return side->ground() && main != side && mainCl != sideCl &&
+    // either they are both induction depth 0 (not yet inducted on)
+    ((!mainCl->inference().inductionDepth() && !sideCl->inference().inductionDepth()) ||
+    // or they are hypothesis and conclusion from the same step and predicates
+    (InductionHelper::isInductionLiteral(side, sideCl) &&
+      InductionHelper::isInductionLiteral(main, mainCl) &&
+      includes(mainSk.begin(), mainSk.end(), sideSk.begin(), sideSk.end()) && !side->isEquality() && !main->isEquality()));
+}
+
+vvector<pair<SLQueryResult, vset<pair<Literal*,Clause*>>>> GeneralInduction::selectMainSidePairs(Literal* literal, Clause* premise)
+{
+  vvector<pair<SLQueryResult, vset<pair<Literal*,Clause*>>>> res;
+  static const bool indmc = env.options->inductionMultiClause();
+
+  // TODO(mhajdu): is there a way to duplicate these iterators?
+  TermQueryResultIterator itSides = TermQueryResultIterator::getEmpty();
+  TermQueryResultIterator itMains = TermQueryResultIterator::getEmpty();
+  if (indmc) {
+    SubtermIterator stit(literal);
+    while (stit.hasNext()) {
+      auto st = stit.next();
+      if (st.isTerm() && skolem(st.term())) {
+        itSides = pvi(getConcatenatedIterator(itSides, _index->getGeneralizations(st)));
+        itMains = pvi(getConcatenatedIterator(itMains, _index->getGeneralizations(st)));
+      }
+    }
+  }
+
+  // pair current literal as main literal with possible side literals
+  // this results in any number of side literals
+  if (InductionHelper::isInductionLiteral(literal))
+  {
+    res.emplace_back(SLQueryResult(literal, premise), vset<pair<Literal*,Clause*>>());
+    while (itSides.hasNext()) {
+      auto qr = itSides.next();
+      if (InductionHelper::isInductionClause(qr.clause) && sideLitCondition(literal, premise, qr.literal, qr.clause)) {
+        res.back().second.emplace(qr.literal, qr.clause);
+      }
+    }
+  }
+  // pair current literal as side literal with possible main literals
+  // this results in only one side literal (the current)
```

Comment:
What about the inference where the current literal is one of multiple side literals? Is there a reason why we don't perform it?

Reply:
The reason is we need another term index query to get other side literals for the main literal which would be as if the main literal triggered the induction. However, we could try it, I guess in some cases it would help, in others it would give more/bigger clauses without any benefit.

Reply:
My point here is that whether we perform the inference then depends on the order of selection, so in some sense we are compromising completeness.
If it is not too much work, could you add that inference and see if it helps or hurts the performance?

Reply:
Done.